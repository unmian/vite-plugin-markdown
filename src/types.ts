/* eslint-disable no-use-before-define */
import type MarkdownIt from 'markdown-it'
import type { FilterPattern } from '@rollup/pluginutils'

export interface Options {
  /**
   * Explicitly set the Vue version
   *
   * @default auto detected
   */
  vueVersion?: string

  /**
   * Enable head support, need to install @vueuse/head and register to App in main.js
   *
   * @default false
   */
  headEnabled?: boolean

  /**
   * The head field in frontmatter used to be used for @vueuse/head
   *
   * When an empty string is passed, it will use the root properties of the frontmatter
   *
   * @default ''
   */
  headField?: string

  /**
   * Parse for frontmatter
   *
   * @default true
   */
  frontmatter?: boolean

  /**
   * Parse for excerpt
   *
   * If `true`, it will be passed to `frontmatterPreprocess` as `frontmatter.excerpt`, replacing the `excerpt` key in frontmatter, if there's any
   *
   * @default false
   */
  excerpt?: boolean

  /**
   * Remove custom SFC block
   *
   * @default ['route', 'i18n']
   */
  customSfcBlocks?: string[]

  /**
   * Custom function to process the frontmatter
   */
  frontmatterPreprocess?: (frontmatter: Record<string, unknown>, options: ResolvedOptions) => any

  /**
   * Expose frontmatter via expose API
   *
   * @default true
   */
  exposeFrontmatter?: boolean

  /**
   * Expose excerpt via expose API
   *
   * @default false
   */
  exposeExcerpt?: boolean

  /**
   * Add `v-pre` to `<code>` tag to escape curly brackets interpolation
   *
   * @see https://github.com/antfu/vite-plugin-mkd/issues/14
   * @default true
   */
  escapeCodeTagInterpolation?: boolean

  /**
   * Options passed to Markdown It
   */
  markdownItOptions?: MarkdownIt.Options

  /**
   * Plugins for Markdown It
   */
  markdownItUses?: (MarkdownIt.PluginSimple | [MarkdownIt.PluginSimple | MarkdownIt.PluginWithOptions<any>, any] | any)[]

  /**
   * A function providing the Markdown It instance gets the ability to apply custom settings/plugins
   */
  markdownItSetup?: (MarkdownIt: MarkdownIt) => void

  /**
   * Options passed to grayMatter
   * TODO grayMatterOptions type
   */
  grayMatterOptions?: any

  /**
   * Class names for wrapper div
   *
   * @default 'markdown-body'
   */
  wrapperClasses?: string | string[]

  /**
   * Component name to wrapper with
   *
   * @default undefined
   */
  wrapperComponent?: string | undefined | null

  /**
   * Custom tranformations apply before and after the markdown transformation
   */
  transforms?: {
    before?: (code: string, id: string) => string
    after?: (code: string, id: string) => string
  }

  include?: FilterPattern
  exclude?: FilterPattern
}

export interface ResolvedOptions extends Required<Options> {
  wrapperClasses: string
}
