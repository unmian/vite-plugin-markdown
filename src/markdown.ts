import MarkdownIt from 'markdown-it'
import matter from 'gray-matter'
import { toArray } from '@antfu/utils'
import type { ResolvedOptions } from './types'

const scriptSetupRE = /<\s*script[^>]*\bsetup\b[^>]*>([\s\S]*)<\/script>/mg
const defineExposeRE = /defineExpose\s*\(/mg
const defineDataRE = /(\/\*[\s\t]*defineData[\s\t]*\*\/[\s\t]*)([A-Za-z0-9]+([\s\t]*,[\s\t]*[A-Za-z0-9]+)*)([\s\n\t;])*/mg
const defineComponentRE = /(\/\*[\s\t]*defineComponent[\s\t]*\*\/[\s\t]*)([A-Za-z0-9]+([\s\t]*,[\s\t]*[A-Za-z0-9]+)*)([\s\n\t;])*/mg
const defineMethodRE = /(\/\*[\s\t]*defineMethod[\s\t]*\*\/[\s\t]*)([A-Za-z0-9]+([\s\t]*,[\s\t]*[A-Za-z0-9]+)*)([\s\n\t;])*/mg

function extractScriptSetup(html: string) {
  const scripts: string[] = []
  html = html.replace(scriptSetupRE, (_, script) => {
    scripts.push(script)
    return ''
  })

  return { html, scripts }
}

function extractCustomBlock(html: string, options: ResolvedOptions) {
  const blocks: string[] = []
  for (const tag of options.customSfcBlocks) {
    html = html.replace(new RegExp(`<${tag}[^>]*\\b[^>]*>[^<>]*<\\/${tag}>`, 'mg'), (code) => {
      blocks.push(code)
      return ''
    })
  }

  return { html, blocks }
}

export function createMarkdown(options: ResolvedOptions) {
  const isVue2 = options.vueVersion.startsWith('2.')

  const markdown = new MarkdownIt({
    html: true,
    linkify: true,
    typographer: true,
    ...options.markdownItOptions,
  })

  markdown.linkify.set({ fuzzyLink: false })

  options.markdownItUses.forEach((e) => {
    const [plugin, options] = toArray(e)

    markdown.use(plugin, options)
  })

  options.markdownItSetup(markdown)

  return (id: string, raw: string) => {
    const { wrapperClasses, wrapperComponent, transforms, headEnabled, frontmatterPreprocess } = options

    raw = raw.trimStart()

    if (transforms.before)
      raw = transforms.before(raw, id)

    if (options.excerpt && !options.grayMatterOptions.excerpt)
      options.grayMatterOptions.excerpt = true

    const grayMatterFile = options.frontmatter
      ? matter(raw, options.grayMatterOptions)
      : { content: raw, data: null, excerpt: '' }
    const { content: md, data } = grayMatterFile
    const excerpt = grayMatterFile.excerpt === undefined ? '' : grayMatterFile.excerpt

    let html = markdown.render(md, { id })

    if (wrapperClasses)
      html = `<div class="${wrapperClasses}">${html}</div>`
    else
      html = `<div>${html}</div>`
    if (wrapperComponent)
      html = `<${wrapperComponent}${options.frontmatter ? ' :frontmatter="frontmatter"' : ''}${options.excerpt ? ' :excerpt="excerpt"' : ''}>${html}</${wrapperComponent}>`
    if (transforms.after)
      html = transforms.after(html, id)

    if (options.escapeCodeTagInterpolation) {
      // escape curly brackets interpolation in <code>, #14
      html = html.replace(/<code(.*?)>/g, '<code$1 v-pre>')
    }

    const hoistScripts = extractScriptSetup(html)
    html = hoistScripts.html
    const customBlocks = extractCustomBlock(html, options)
    html = customBlocks.html

    const customDataBlocks: string[] = []
    const customComponentBlocks: string[] = []
    const customMethodBlocks: string[] = []
    if (isVue2) {
      hoistScripts.scripts = hoistScripts.scripts.map((line) => {
        line = line.replace(defineDataRE, (string, p1, p2) => {
          customDataBlocks.push(p2.replace(/[\s\t]+/g, ''))
          return string.replace(p1, '')
        })
        line = line.replace(defineComponentRE, (string, p1, p2) => {
          customComponentBlocks.push(p2.replace(/[\s\t]+/g, ''))
          return string.replace(p1, '')
        })
        line = line.replace(defineMethodRE, (string, p1, p2) => {
          customMethodBlocks.push(p2.replace(/[\s\t]+/g, ''))
          return string.replace(p1, '')
        })
        return line
      })
    }

    const scriptLines: string[] = []
    let frontmatterExportsLines: string[] = []
    let excerptExportsLine = ''
    let excerptKeyOverlapping = false

    if (options.frontmatter) {
      if (options.excerpt && data) {
        if (data.excerpt !== undefined) excerptKeyOverlapping = true
        data.excerpt = excerpt
      }

      const { head, frontmatter } = frontmatterPreprocess(data || {}, options)

      if (options.excerpt && !excerptKeyOverlapping && frontmatter.excerpt !== undefined)
        delete frontmatter.excerpt

      scriptLines.push(`const frontmatter = ${JSON.stringify(frontmatter)}`)

      frontmatterExportsLines = Object.entries(frontmatter).map(([key, value]) => `export const ${key} = ${JSON.stringify(value)}`)

      if (!isVue2 && options.exposeFrontmatter && !defineExposeRE.test(hoistScripts.scripts.join('')))
        scriptLines.push('defineExpose({ frontmatter })')

      if (!isVue2 && headEnabled && head) {
        scriptLines.push(`const head = ${JSON.stringify(head)}`)
        scriptLines.unshift('import { useHead } from "@vueuse/head"')
        scriptLines.push('useHead(head)')
      }
    }

    if (options.excerpt) {
      scriptLines.push(`const excerpt = ${JSON.stringify(excerpt)}`)

      if (!excerptKeyOverlapping) excerptExportsLine = `export const excerpt = ${JSON.stringify(excerpt)}\n`

      if (!isVue2 && options.exposeExcerpt && !defineExposeRE.test(hoistScripts.scripts.join('')))
        scriptLines.push('defineExpose({ excerpt })')
    }

    scriptLines.push(...hoistScripts.scripts)

    const scripts = isVue2
      ? `<script>\n${scriptLines.join('\n')}\n${frontmatterExportsLines.join('\n')}\n${excerptExportsLine}export default { components: { ${customComponentBlocks.join(',')} },data() { return { frontmatter,${customDataBlocks.join(',')} } }, methods: { ${customMethodBlocks.join(',')} } }\n</script>`
      : `<script setup>\n${scriptLines.join('\n')}\n</script>${frontmatterExportsLines.length ? `\n<script>\n${frontmatterExportsLines.join('\n')}\n</script>` : ''}${excerptExportsLine !== '' ? `\n<script>\n${excerptExportsLine}</script>` : ''}`

    const sfc = `<template>${html}</template>\n${scripts}\n${customBlocks.blocks.join('\n')}\n`

    return sfc
  }
}
