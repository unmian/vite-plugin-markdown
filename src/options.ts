import { toArray } from '@antfu/utils'
import { preprocessHead } from './head'
import type { Options, ResolvedOptions } from './types'
import { getVueVersion } from './utils'

export function resolveOptions(userOptions: Options): ResolvedOptions {
  const options = Object.assign({
    headEnabled: false,
    headField: '',
    frontmatter: true,
    excerpt: false,
    exposeFrontmatter: true,
    exposeExcerpt: false,
    escapeCodeTagInterpolation: true,
    customSfcBlocks: ['route', 'i18n', 'style'],
    markdownItOptions: {},
    markdownItUses: [],
    markdownItSetup: () => {},
    grayMatterOptions: {},
    wrapperClasses: 'markdown-body',
    wrapperComponent: null,
    transforms: {},
    frontmatterPreprocess: (frontmatter: any, options: ResolvedOptions) => {
      const head = preprocessHead(frontmatter, options)
      return { head, frontmatter }
    },
  }, userOptions) as ResolvedOptions

  options.wrapperClasses = toArray(options.wrapperClasses).filter(i => i).join(' ')
  options.vueVersion = options.vueVersion || getVueVersion()

  return options
}
