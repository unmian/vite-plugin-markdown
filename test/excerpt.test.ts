import { describe, expect, it } from 'vitest'
import { createMarkdown } from '../src/markdown'
import { resolveOptions } from '../src/options'

describe('excerpt', () => {
  const options = resolveOptions({
    excerpt: true,
    grayMatterOptions: {
      excerpt: true,
      excerpt_separator: '<!-- more -->',
    },
  })
  const markdownToVue = createMarkdown(options)

  it('basic-excerpt', () => {
    const md = `---
title: Hey
---

This is an excerpt.

<!-- more -->

# Hello

- A
- B
- C`
    expect(markdownToVue('', md)).toMatchSnapshot()
  })
})
