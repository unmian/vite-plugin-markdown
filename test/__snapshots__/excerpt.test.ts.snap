// Vitest Snapshot v1

exports[`excerpt > basic-excerpt 1`] = `
"<template><div class=\\"markdown-body\\"><p>This is an excerpt.</p>
<!-- more -->
<h1>Hello</h1>
<ul>
<li>A</li>
<li>B</li>
<li>C</li>
</ul>
</div></template>
<script setup>
const frontmatter = {\\"title\\":\\"Hey\\"}
defineExpose({ frontmatter })
const excerpt = \\"\\\\nThis is an excerpt.\\\\n\\\\n\\"
</script>
<script>
export const title = \\"Hey\\"
</script>
<script>
export const excerpt = \\"\\\\nThis is an excerpt.\\\\n\\\\n\\"
</script>

"
`;
